viewing_party/party.py

```diff
@@ -1,23 +1,170 @@
+                and movie not in friends_unique_movies:
+                friends_unique_movies.append(movie)
+
+    return friends_unique_movies
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    recommended_movies_list = []
+    user_watched_titles = get_user_watched_list(user_data)
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched_titles \
+                and movie["host"] in user_data["subscriptions"] \
+                and movie not in recommended_movies_list:
+                    recommended_movies_list.append(movie)
+
+    return recommended_movies_list
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    recs = []
+    if not user_data["watched"]:
+        return recs   
+    if not any(friend["watched"] for friend in user_data["friends"]):
+        return recs  # Return empty list since there are no watched movies by friends  
+    
+    fav_genre = get_favorite_genre(user_data)
+    user_watched_titles = get_user_watched_list(user_data)
+    friend_watched = get_friends_watched_list_unique(user_data)
+    
+    for movie in friend_watched:
+        if (movie["title"] not in user_watched_titles \
+            and movie["genre"] == fav_genre \
+            and movie not in recs):
+            recs.append(movie)
+    return recs
+ 
+def get_rec_from_favorites(user_data):
+    recs = []
+    friend_watched = get_friends_watched_list_unique(user_data)
+    friends_watched_titles = []
+    
+    for movie in friend_watched:
+        friends_watched_titles.append(movie["title"])
+   
+    for movie in user_data["favorites"]:
+        if movie["title"] not in friends_watched_titles:
+                recs.append(movie)
+    return recs
+
+#helper function
+def get_friends_watched_list_unique(user_data):
+    friends_watched = []
+    friends_watched_titles = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]: 
+            if movie["title"] not in friends_watched_titles:
+                friends_watched_titles.add(movie["title"])            
+                friends_watched.append(movie)
+    return friends_watched
+
+#helper function
+def get_favorite_genre(user_data):
+    genre_freq= {}
+    fav_genre = None
+    max_count = 0  
+    for movie in user_data["watched"]:   
+        genre = movie["genre"]
+        genre_freq[genre] = genre_freq.get(genre,0) + 1
+        if genre_freq[genre] > max_count:
+            max_count = genre_freq[genre]
+            fav_genre =genre
+    return fav_genre
 
+# Assumed user_watched_list helper function
+def get_user_watched_list(user_data):
+    return [movie["title"] for movie in user_data["watched"]]
```

Comment:
Love the helper functions! Now that we know about imports, could we move them to a different file and import them? 👀